### PR TITLE
added initial reward collision model

### PR DESCRIPTION
**Pull request by oravner**

### CODE REVIEW

File: pre_trained_reward.py [hunk 1]
```diff
@@ -24,7 +24,7 @@ def __init__(self, model_name, config):
         if self.is_vision_enabled:
             self.workspace_image_inputs = tf.placeholder(tf.float32, (None, 55, 111), name='workspace_image_inputs')
             self.images_3d = tf.expand_dims(self.workspace_image_inputs, axis=-1)
-        self.goal_pose_inputs = tf.placeholder(tf.float32, (None, 2), name='goal_pose_inputs')
+        self.goal_pose_inputs = None # tf.placeholder(tf.float32, (None, 2), name='goal_pose_inputs')
```
Comment (tomjur) [inline pre_trained_reward.py]: please remove unused inputs - you will also need to change make feed, remove _generate_goal_features and anything else that is no longer relevant

File: pre_trained_reward.py [hunk 2]
```diff
@@ -182,7 +182,7 @@ def get_batch_and_labels(batch, openrave_manager, image_cache):
             start_joints, goal_joints, action, next_joints, reward, terminated, status = batch[i]
         else:
             workspace_id, start_joints, goal_joints, action, next_joints, reward, terminated, status = batch[i]
-        goal_pose = openrave_manager.get_target_pose(goal_joints)
+        goal_pose = None #openrave_manager.get_target_pose(goal_joints)
```
Comment (tomjur) [inline pre_trained_reward.py]: remove goal_pose entirely + remove the newly unused openrave_manager -> see if you can also remove the dependency on OpenRaveManager completely for the training process

File: reward_collision_network.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+                layer_size,
+                activation=_activation,
+                name='{}_layers_{}'.format(name_prefix, i),
+                kernel_regularizer=tf_layers.l2_regularizer(l2_scale),
+                reuse=self._reuse_flag
+            )
+        softmax_logits = current
+
+        # TODO: add joints_clipping loss?
+
```
Comment (tomjur) [inline reward_collision_network.py]: can remove the TODO (no need to add)

File: reward_collision_model.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
+from reward_collision_network import CollisionNetwork
```
Comment (tomjur) [inline reward_collision_model.py:1]: general comments for file:
1. why do we need the class TensorBoard and the logic is not inside CollisionModel? it would be less messy to have it inside CollisionModel - or do we need it elsewhere also?
2. this file is a trainer class right? if so please rename CollisionModel to CollisionTrainer

File: reward_collision_model.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        self.optimizer = self.init_optimizer()
+
+        with open(os.path.join(self.model_dir, 'config.yml'), 'w') as fd:
+            yaml.dump(config, fd)
+
+        self.train_board = TensorBoard(tensorboard_dir, 'train_' + model_name, self.train_summaries)
+        self.test_board = TensorBoard(tensorboard_dir, 'test_' + model_name, self.test_summaries)
+
+    def predict(self, data_batch, session):
+        feed = self.network.make_feed(*data_batch)
```
Comment (tomjur) [inline reward_collision_model.py]: i think it would be clearer to partition data_batch to the actual inputs it should take.
this way the client code can see which parameters must be passed to the model

File: reward_collision_model.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+        self.epochs = config['general']['epochs']
+        self.save_every_epochs = config['general']['save_every_epochs']
+
+        self.network = CollisionNetwork(config, self.model_dir)
+        self.net_output = self.network.status_softmax_logits
+        self.status_input = tf.placeholder(tf.int32, [None, ])
+        self.prediction = tf.argmax(tf.nn.softmax(self.net_output), axis=1)
+
+        self.global_step = 0
```
Comment (tomjur) [inline reward_collision_model.py:36]: i think that if you don't use a global_step var, then tf keeps a global step var that gets updated every optimize call.
summaries are automatically updated accordingly

File: reward_collision_model.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+
+        gradients, variables = zip(*optimizer.compute_gradients(self.loss, tf.trainable_variables()))
+        initial_gradients_norm = tf.global_norm(gradients)
+        gradient_limit = config['reward']['gradient_limit']
+        if gradient_limit > 0.0:
+            gradients, _ = tf.clip_by_global_norm(gradients, gradient_limit, use_norm=initial_gradients_norm)
+        clipped_gradients_norm = tf.global_norm(gradients)
+        self.train_summaries.append(tf.summary.scalar('Gradients_Norm_Initial', initial_gradients_norm))
+        self.train_summaries.append(tf.summary.scalar('Gradients_Norm_Clipped', clipped_gradients_norm))
+
```
Comment (tomjur) [inline reward_collision_model.py:116]: please also add the gradient summaries to the testing phase, since we also want to make sure that the gradient behave well during testing (:

File: reward_collision_model.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        self.train_board.writer.add_summary(train_summary, self.global_step)
+
+    def _test_batch(self, test_batch, status_batch, session):
+        test_feed = self.network.make_feed(*test_batch)
+        test_feed[self.status_input] = np.array(status_batch)
+        test_summary = session.run(
+            [self.test_board.summaries] + self.test_measures,
+            test_feed)[0]
+        self.test_board.writer.add_summary(test_summary, self.global_step)
+        self.test_board.writer.flush()
```
Comment (tomjur) [inline reward_collision_model.py:136]: for both methods above see comments about:
global step
data_batch -> explicit parameters

File: reward_collision_model.py [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    def add_test_measures(self):
+        labels = self.status_input - 1
+        accuracy, accuracy_op = tf.metrics.accuracy(labels=labels, predictions=self.prediction)
+        precision, precision_op = tf.metrics.precision(labels=labels, predictions=self.prediction)
+        recall, recall_op = tf.metrics.recall(labels=labels, predictions=self.prediction)
+        accuracy_summary = tf.summary.scalar('Accuracy', accuracy)
+        recall_summary = tf.summary.scalar('Recall', recall)
+        precision_summary = tf.summary.scalar('Precision', precision)
+        self.test_summaries += [accuracy_summary, recall_summary, precision_summary]
+        return [accuracy, accuracy_op, recall, recall_op, precision, precision_op]
```
Comment (tomjur) [inline reward_collision_model.py]: are you calling these ops? or just using them inside tensor board? if just inside TB, no need to return

File: reward_collision_model.py [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+        test_feed[self.status_input] = np.array(status_batch)
+        test_summary = session.run(
+            [self.test_board.summaries] + self.test_measures,
+            test_feed)[0]
+        self.test_board.writer.add_summary(test_summary, self.global_step)
+        self.test_board.writer.flush()
+
+    def train(self, train_data, test_data, image_cache, session):
+        session.run(tf.global_variables_initializer())
+        session.run(tf.local_variables_initializer())
```
Comment (tomjur) [inline reward_collision_model.py:140]: tf.initialize_all_variables()

File: reward_collision_model.py [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+
+        for epoch in range(self.epochs):
+
+            for train_batch in train_data:
+                train_batch, train_status_batch = get_batch_and_labels(train_batch, image_cache)
+                # TODO: assert if train_status contains goal status
```
Comment (tomjur) [inline reward_collision_model.py]: what does this comment mean?

File: reward_data_manager.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
+import time
```
Comment (tomjur) [inline reward_data_manager.py:1]: do we really need everything here?

**Review by tomjur — CHANGES_REQUESTED** [review]
great job - definitely more readable than my previous main function (:
 see comments and questions inline!

File: pre_trained_reward.py (hunk 2 above)
Comment (oravner) [reply pre_trained_reward.py]: Sorry I forgot to remove this file. Please notice that it breaks the whole project since we haven't wrote the deterministic reward model which wraps the collision model.

File: reward_collision_network.py (hunk 3 above)
Comment (oravner) [reply reward_collision_network.py]: V

File: reward_collision_model.py (hunk 4 above)
Comment (oravner) [reply reward_collision_model.py:1]: 1. I had many dilemmas how to use the tensorboard reporting since it's API is not very design convenient. Good idea, it will be an internal class inside the CollisionModel. If we will use it in other models as well (such as the Reward itself) I will move it to some utility file.
2. I thought to rather separate the 2 models to: 1. Model which only holds the network logic and architecture. 2. Model which wraps the network logic for training and prediction. What do you think is better and why?

File: reward_collision_model.py (hunk 5 above)
Comment (oravner) [reply reward_collision_model.py]: V

File: pre_trained_reward.py (hunk 1 above)
Comment (oravner) [reply pre_trained_reward.py]: Sorry I forgot to remove this file. Please notice that it breaks the whole project since we haven't wrote the deterministic reward model which wraps the collision model.

File: reward_collision_model.py (hunk 6 above)
Comment (oravner) [reply reward_collision_model.py:36]: I don't understand. You recommend removing this var and leave only the tf Variable? This is the same logic as was in the reward model code.

File: reward_collision_model.py (hunk 7 above)
Comment (oravner) [reply reward_collision_model.py:116]: V

File: reward_collision_model.py (hunk 9 above)
Comment (oravner) [reply reward_collision_model.py]: The session must run them

File: reward_collision_model.py (hunk 10 above)
Comment (oravner) [reply reward_collision_model.py:140]: I got warning tf.initialize_all_variables() is deprecated and should use tf.global_variables_initializer() instead ):

File: reward_collision_model.py (hunk 11 above)
Comment (oravner) [reply reward_collision_model.py]: I assumed that all the data doesn't contain any goal status sample, so it will be best practice to assert here if it does exist

File: reward_data_manager.py (hunk 12 above)
Comment (oravner) [reply reward_data_manager.py:1]: only the datetime was unnecessary ):

File: pre_trained_reward.py (hunk 1 above)
Comment (tomjur) [reply pre_trained_reward.py]: no problem, the next step should fix this.
i just don't want to carry over dead code that we'll forget to delete

File: reward_collision_model.py (hunk 4 above)
Comment (tomjur) [reply reward_collision_model.py:1]: I think you have it right - 2 classes work best. what I meant in my comment was to just to rename to "Trainer" since the role of the non-network class is to train the network class.

File: reward_collision_model.py (hunk 6 above)
Comment (tomjur) [reply reward_collision_model.py:36]: maybe the reward model was also incorrect (:

I think that if you don't specify a global step (ever) than tf holds an internal counter that raises every time you call an optimize function.

maybe we can remove and see if everything just works as expected?

File: reward_collision_model.py (hunk 9 above)
Comment (tomjur) [reply reward_collision_model.py]: ok

File: reward_collision_model.py (hunk 10 above)
Comment (tomjur) [reply reward_collision_model.py:140]: so why the second line? also a warning?
this comment shows i was working with a deprecated tf version (:

File: reward_collision_model.py (hunk 11 above)
Comment (tomjur) [reply reward_collision_model.py]: ok, so use assert so it is not left as a comment

File: reward_collision_model.py (hunk 6 above)
Comment (oravner) [reply reward_collision_model.py:36]: according to the tf doc, i need to create tf.variable of global_step and pass it to minimize / apply_gradients for it to be updated. Then I need to create my own pythonic variable of global step to pass to tensorboard and the tf.saver

File: reward_collision_model.py (hunk 10 above)
Comment (oravner) [reply reward_collision_model.py:140]: for some reason some variables are classified as globals and some are locals

File: reward_collision_model.py (hunk 11 above)
Comment (oravner) [reply reward_collision_model.py]: V

File: data/config/reward_config.yml [hunk 13]
```diff
@@ -1,14 +1,14 @@
 general:
```
Comment (tomjur) [inline data/config/reward_config.yml:1]: are all the below changes justified? especially the oversampling ratios...

File: reward_collision_model.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+        self.test_summaries = []
+
+        self.epochs = config['general']['epochs']
+        self.save_every_epochs = config['general']['save_every_epochs']
+
+        self.network = CollisionNetwork(config, self.model_dir)
+        self.net_output = self.network.status_softmax_logits
+        self.status_input = tf.placeholder(tf.int32, [None, ])
+        # TODO: fix collision_prob
+        self.collision_prob = tf.expand_dims(tf.nn.softmax(self.net_output)[:, 1], -1)
```
Comment (tomjur) [inline reward_collision_model.py:33]: todo-?
